### ISSUE TITLE
Update to React 16

### ISSUE THREAD

**Issue by vam** [crudlio/crudl#51]

**Comment by jpic** [comment]
i'm looking to make new friends to try and pair program to port Crudl to Material-UI + Next.JS

**Comment by jpic** [comment]
It turns out our frontend team already has experience with ant.design and I didn't know about it, so we'll reconsider that.

PouchDB: Ok we will make the connector an external package.

Nextjs: i understand your point, but in django it's been such a misery that all features were added to the admin and that we always had to start from scratch when building user facing cruds. I don't want to fall in the same trap, and have a crud framework that can at least do SSR for list and detail pages that can be open to public. But if that doesn't work well 

Thanks for the heads up, really useful !

**Comment by sehmaschine** [comment]
About Next: I fully agree with what you're saying about Django. But with regards to CRUDL it comes down to Next vs React and I don't quite see the benefits here.

**Comment by jpic** [comment]
What do you mean Next vs. React ? Next depends on react and react-dom it's
fully React oriented.

**Comment by sehmaschine** [comment]
I thought your idea is to use Next instead of React with CRUDL (I know that Next is based on React, but it still means switching the technology). Isn't that your proposal? Did I miss someting?

**Comment by jpic** [comment]
The proposal is rather changing from gulp to webpack, but instead of
removing gulpfile and adding webpack.config, just remove gulpfile, and have
this work by running command "next" because nextjs bundles a pretty cool
webpack config, with hot module reload, ssr, link prefetching ...

**Comment by sehmaschine** [comment]
Now I see. We planned to use create-react-app in order to go with webpack.

**Comment by jpic** [comment]
Yes it's either create-react-app either nextjs, but you have to go through
quite some steps to have SSR:
https://medium.com/bucharestjs/upgrading-a-create-react-app-project-to-a-ssr-code-splitting-setup-9da57df2040a

So, if you're bundling crudlio as dependency in your project bundle you
have to go through this again. I'm open on CRA, Next, or other, but for me
SSR for me is a must, as is content prefechting which i don't know if CRA
supports, but Next does for sure with the Link component.

That said, I need to try CRA to compare, probably you want to try NextJS
too (it's pretty amazing how it gets you started in minutes !).

**Comment by sehmaschine** [comment]
I already tried Next and I do like it. As said before, I'm just not convinced (yet) that it has benefits (e.g. compared to CRA). What I don't like about Next is the strict folder structure.

But I'm not so sure if we agree on _why_ the switch to either CRA/Next is necessary.
1. You mention SSR and I don't quite get why this is useful with CRUDL!?
2. You say "bundling crudlio as a dependency". Why do you want to do that (because implementing CRUDL usually means to use the js from the CDN)?

Just a guess: Do you plan to use CRUDL _within_ one of your projects where you need SSR (not because of CRUDL, but because of the rest of your code)?

**Comment by jpic** [comment]
1. I think SSR is a requirement because I want to reuse crudl stuff in public frontend, bots to index them properly, link sharing to work as expected by the internet, who's not evaluating javascript every time they fetch a page programatically after the user provided a link (even though personnaly I have a [receipe for this with fetch+jsdom](https://github.com/betagouv/mrs/blob/master/src/mrs/static/js/smoke.test.js)). For me not having SSR feels like only having ipv6 if that makes any sense ^^

2. When we make a new project, we will likely want to add more than crudl views, like new Forms, or Actions (that can be executed on a single object or multiple objects (from the list view)). In "another project", I want to be able to import the crud router, and redefine it for a particular model, add views, forms, actions to it. I expect the crud to provide a nice UI by default, allow tweaking variables such as table columns etc, but oalso to be able to completely override the crud object which means importing it, extending it, and registering it to crudl.

So really, i'm not looking for a new django admin app, but rather a new framework that provides with modern CRUD experience OOTB. Does that make sense to you ?

**Comment by sehmaschine** [comment]
I see what you mean (at least I think so).

+1 for overridable, reusable.
I'd say go ahead with the fork and we'll see what we finally use with CRUDL. I still have my doubts about Next, but that doesn't mean switching is unlikely. In general, anything which makes CRUDL more customizable (and easier to maintain) is appreciated.
